### PR TITLE
Introduce BaseSpec class for test suites

### PR DESCRIPTION
**Pull request by srstevenson**
This PR extracts a `BaseSpec` class containing common test fixtures, to reduce test code duplication.

### CODE REVIEW

File: src/test/scala/com/asidatascience/configuration/BaseSpec.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.concurrent.atomic.AtomicInteger
+import scala.concurrent.duration._
+import scala.util.{Success, Try}
+
+import akka.actor.ActorSystem
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}
+
+trait BaseSpec
+extends FlatSpec
```
Comment (pbugnion) [inline src/test/scala/com/asidatascience/configuration/BaseSpec.scala]: Isn't FlatSpec a class? Shouldn't BaseSpec be an abstract class?

Comment (pbugnion) [inline src/test/scala/com/asidatascience/configuration/BaseSpec.scala]: See e.g. http://www.scalatest.org/user_guide/defining_base_classes

**Comment by pbugnion** [comment]
This looks good, modulo comments on trait vs class.

File: src/test/scala/com/asidatascience/configuration/BaseSpec.scala (hunk 1 above)
Comment (srstevenson) [inline src/test/scala/com/asidatascience/configuration/BaseSpec.scala]: I'm not sure, I'm happy to defer judgement. Using a trait was my default choice, and the compiler was happy with it. Reading around (e.g. [1](http://alvinalexander.com/scala/scala-trait-examples#limiting-which-classes-can-use-a-trait), [2](https://stackoverflow.com/questions/12854941/why-can-a-scala-trait-extend-a-class)), it looks to be discouraged though (and if `BaseSpec` is a trait extending `FlatSpec`, is it missing a self type `this: FlatSpec =>`?).

Comment (pbugnion) [inline src/test/scala/com/asidatascience/configuration/BaseSpec.scala]: Interesting. I didn't know traits could extend a class. It looks like the benefits of doing this are smilar to the benefits of using a self type. Ie.

```scala
class FlatSpec

trait BaseSpec extends FlatSpec
```

is similar to:

```scala
class FlatSpec

trait BaseSpec {
  this: FlatSpec =>
}
```
inasmuch as you're forcing subclasses of `BaseSpec` to also be subclasses of `FlatSpec`.

Anyway, we can avoid all this complicated typing by just using an abstract class:

```scala
abstract class BaseSpec extends FlatSpec
```

This is definitely what scalatest recommends.

**Comment by srstevenson** [comment]
I've changed `BaseSpec` to a class.

**Comment by pbugnion** [comment]
I think this is good!

**Comment by pbugnion** [comment]
We can't actually merge this until after PR #19 .

**Comment by srstevenson** [comment]
@pbugnion: I've rebased this on #19.

**Pull request merged**